File: paplj/src/main/java/com/virtlink/paplj/DummyCodeCompleter.java
package com.virtlink.paplj;

import com.virtlink.editorservices.*;
import com.virtlink.editorservices.codecompletion.*;
import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;
import java.util.EnumSet;
import java.util.List;

public class DummyCodeCompleter implements ICodeCompleter {

    @NotNull
    @Override
    public ICompletionInfo complete(IDocument iDocument, int i, ICancellationToken iCancellationToken) {
        List<ICompletionProposal> proposals = new ArrayList<>();
        proposals.add(new CompletionProposal("Proposal1", "Description", "Documentation", false, "insertion1", 0, null, null, 0, Kind.Method, new Visibility(ClassVisibility.Private, PackageVisibility.Package, LibraryVisibility.Internal), EnumSet.of(Attribute.Static)));
        proposals.add(new CompletionProposal("Proposal 2 (another prposal)", "Description 2", "Documentation 2", false, "another2", 0, null, null, 0, Kind.Method, new Visibility(ClassVisibility.Private, PackageVisibility.Package, LibraryVisibility.Internal), EnumSet.of(Attribute.Static)));

        // TODO: Determine prefix according to language rules.
        return new CompletionInfo("Pro", proposals);
    }
}

File: paplj/src/main/kotlin/com/virtlink/paplj/codecompletion/DummyCodeCompleter.kt
package com.virtlink.paplj.codecompletion

import com.virtlink.editorservices.*
import com.virtlink.editorservices.codecompletion.*
import com.virtlink.logging.logger
import java.net.URI

class DummyCodeCompleter : ICodeCompletionService {

    override val triggerCharacters: List<Char>
        get() = emptyList()

    @Suppress("PrivatePropertyName")
    private val LOG by logger()

    override fun configure(configuration: ICodeCompletionConfiguration) {
        // Nothing to do.
    }

    override fun getCompletionInfo(document: URI, caretOffset: Offset, cancellationToken: ICancellationToken?): ICompletionInfo {
        LOG.info("$document: Completing at $caretOffset.")

        val proposals = listOf(
                CompletionProposal("Hello",
                        description = "Description string",
                        content = "hello world!",
                        scopes = "meta.field,meta.static"),
                CompletionProposal("Local variable",
                        content = "local var",
                        scopes = "meta.variable,meta.internal"),
                CompletionProposal("Method",
                        content = "method()",
                        scopes = "meta.method,meta.abstract,meta.deprecated,meta.package"),
                CompletionProposal("if (then else)")
        )

        // TODO: Determine prefix according to language rules.
        return CompletionInfo("", proposals)
    }
}